rfcs/rfc-3-layout-callbacks.md

```diff
@@ -0,0 +1,222 @@
+
+```rust
+trait WidgetProps: /* ... */ {
+  // ...
+  fn get_on_layout(&self) -> Option<OnLayout>;
+}
+```
+
+Which could be generated by the derive helper macro:
+
+```rust
+#[derive(WidgetProps, Default, Debug, Clone, PartialEq)]
+struct MyWidgetProps {
+  #[prop_field(OnLayout)]
+  on_layout: Option<OnLayout>
+}
+```
+
+And easily accessed by widget via;
+
+```rust
+trait WidgetProps: /* ... */ {
+  // ...
+  fn on_layout(&mut self, context: &mut KayakContextRef, event: &mut LayoutEvent);
+}
+```
+
+##### Considerations
+
+Of course, it's important to consider that this makes the `on_layout` prop completely optional. If a third party widget crate created, they will need to add this field to their widgets in order for users to access the event.
+
+An alternative to this would be to put `on_layout` on `Widget` directly. However, this becomes messy because we'd need to parse that prop distinctly from the others. Additionally, layout events are worthless on widgets that don't have a layout (i.e., they're `RenderCommand::Empty`).
+
+For these reasons, it makes the most sense to put it on `WidgetProps` and let the user opt-in to them, despite it possibly being a bit annoying for library authors.
+
+#### Dispatching the Event
+
+##### Management
+
+The big question for this feature is *who's in charge of the event?* Obviously, it's an internal thing that users and integration crates won't need to manage themselves. But should the responsibility fall on `WidgetManager` or `KayakContext`? And should it go through `EventDispatcher`, a custom dispatcher, or just a simple function?
+
+While it would make sense to store this on `WidgetManager`, it's not possible due to the requirement that we pass a mutable `KayakContext` to the layout event. This means the dispatching must take place within `KayakContext`. 
+
+As for which dispatcher to use, it might be worth creating a custom `LayoutEventDispatcher` to handle the layout events. The reason for this is purely organizational. It's probably better to separate out the concerns so that `KayakContext` doesn't get too bloated.
+
+##### Logic
+
+One complication with the logic is determining which widgets to notify. To keep it simple, we can just notify all widgets that have been re-rendered. 
+
+Another additional complication is that we can't just dispatch the event over all indices in `WidgetManager::dirty_nodes` since those only contain widgets that requested a re-render (i.e., their state updated or a binding marked them dirty). Iterating over only those would skip out on all the children, which may have had their own layouts changed after rendering.
+
+This means we actually need to use `WidgetManager::dirty_render_nodes` in order to get all re-rendered nodes. So instead of draining this collection in `WidgetManager::render()`, we'll need to keep them around and drain them after calculating the layout in `KayakContext::render()`— which we'll do in `LayoutEventDispatcher::dispatch(...)`.
```

Comment:
Layout calculation doesn't happen until morphorm is called. The entire tree is recalculated every frame. We would need to call `on_layout` for all widgets after morphorm runs?

Reply:
I don't think we need to do it for all widgets every frame. We just need to do it for widgets that have been re-rendered (marked dirty).

So the first render will have all widgets, but subsequent ones should only concern the widgets that actually changed— assuming only updated widgets are marked as "dirty render nodes".

Reply:
So we would need to not drain the dirty nodes and instead clear them after we call the `on_layout` after morphorm runs?

Reply:
Yeah I think so 🤔

Reply:
What about children of dirty nodes? Or even parents that might have changed because a child re-rendered?

Reply:
The children seem to also be added to dirty render nodes (but not dirty nodes) so they should be fine. I'll run some more tests to ensure that's always the case.

But you raise a fair point about the parent widget. We probably need to add the parents of each sub-tree to the list of nodes to dispatch, which likely only requires we iterate once through the nodes to add all missing parents (it's an `IndexSet` so it will automatically prevent duplicates) before draining and dispatching.

Reply:
> The children seem to also be added to dirty render nodes (but not dirty nodes) so they should be fine. I'll run some more tests to ensure that's always the case.

Running the `Counter` example, it seems like every widget in the tree within the updated widget is properly added to `WidgetManager::dirty_render_nodes`. So we should be okay to iterate through that when firing the layout events.

> But you raise a fair point about the parent widget. We probably need to add the parents of each sub-tree to the list of nodes to dispatch, which likely only requires we iterate once through the nodes to add all missing parents (it's an `IndexSet` so it will automatically prevent duplicates) before draining and dispatching.

One other point to make about this, is we'll probably want to snag the _valid_ parent via `WidgetManager::get_valid_parent(...)` instead of from the normal tree.

We could try to optimize by just adding the parent of the last widget in the set (since it's an ordered `IndexSet`, the last entry is the root of the subtree). This would save a lookup on each node in the subtree. However, if we did do this, we'd need to keep in mind provider/consumer contexts. Right now, if the data is changed by a descendant, both the provider and all its children are re-rendered. But this may not always be the case— in the future we might update only the consumers rather than the whole subtree below the provider.

So is that an optimization we need/want? Or maybe we just leave that decision until after implementation when we can actually bench test and compare results. Any thoughts? 😅

Reply:
Lets wait for now.